File: algoritmos/WinogradOriginal/src/c/WinogradOriginal.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <string.h>
#include <errno.h>

// Implementación de la multiplicación de matrices con el algoritmo Winograd
void multiplyMatricesWinograd(long long **a, long long **b, long long **result, int rowA, int colA, int colB) {
    long long *rowFactor = calloc(rowA, sizeof(long long));
    long long *colFactor = calloc(colB, sizeof(long long));

    for (int i = 0; i < rowA; i++)
        for (int j = 0; j < colA / 2; j++)
            rowFactor[i] += a[i][2 * j] * a[i][2 * j + 1];

    for (int i = 0; i < colB; i++)
        for (int j = 0; j < rowA / 2; j++)
            colFactor[i] += b[2 * j][i] * b[2 * j + 1][i];

    for (int i = 0; i < rowA; i++) {
        for (int j = 0; j < colB; j++) {
            result[i][j] = -rowFactor[i] - colFactor[j];
            for (int k = 0; k < colA / 2; k++) {
                result[i][j] += (a[i][2 * k] + b[2 * k + 1][j]) *
                                (a[i][2 * k + 1] + b[2 * k][j]);
            }
        }
    }

    if (colA % 2 == 1) {
        for (int i = 0; i < rowA; i++)
            for (int j = 0; j < colB; j++)
                result[i][j] += a[i][colA - 1] * b[colA - 1][j];
    }

    free(rowFactor);
    free(colFactor);
}

// Función para leer matrices desde un archivo
void readMatrices(const char *filename, long long ****matrices, int **rows, int **cols, int *matrixCount) {
    FILE *file = fopen(filename, "r");
    if (!file) {
        perror("No se puede abrir el archivo");
        exit(EXIT_FAILURE);
    }

    long long ***tempMatrices = malloc(10 * sizeof(long long **));
    *rows = malloc(10 * sizeof(int));
    *cols = malloc(10 * sizeof(int));
    int count = 0;

    while (!feof(file)) {
        long long **matrix = NULL;
        int r = 0, c = 0;
        char buffer[8192]; // Tamaño aumentado para grandes matrices

        while (fgets(buffer, sizeof(buffer), file)) {
            if (buffer[0] == '\n') {
                if (r > 0) break;
                continue;
            }

            int colCount = 0;
            char *token = strtok(buffer, " ");
            while (token != NULL) {
                if (colCount == 0) {
                    matrix = realloc(matrix, sizeof(long long *) * (r + 1));
                    if (!matrix) {
                        perror("Error al reasignar memoria");
                        exit(EXIT_FAILURE);
                    }
                    matrix[r] = malloc(sizeof(long long) * (strlen(buffer) / 2 + 1));
                }
                matrix[r][colCount++] = atoll(token);
                token = strtok(NULL, " ");
            }

            if (r == 0) {
                c = colCount;
            } else if (colCount != c) {
                fprintf(stderr, "Error: todas las filas deben tener el mismo número de columnas\n");
                exit(EXIT_FAILURE);
            }

            r++;
        }

        if (r > 0) {
            tempMatrices[count] = matrix;
            (*rows)[count] = r;
            (*cols)[count] = c;
            count++;
        }
    }

    fclose(file);
    *matrices = realloc(tempMatrices, sizeof(long long **) * count);
    *matrixCount = count;
}

// Función para escribir la matriz de resultados en un archivo
void writeMatrix(const char *filename, long long **matrix, int rows, int cols) {
    FILE *file = fopen(filename, "w");
    if (!file) {
        perror("No se puede crear el archivo");
        exit(EXIT_FAILURE);
    }

    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            fprintf(file, "%lld ", matrix[i][j]);
        }
        fprintf(file, "\n");
    }

    fclose(file);
}

// Función para registrar el tiempo de ejecución
void writeExecutionTime(const char *filename, double time) {
    FILE *file = fopen(filename, "w");
    if (!file) {
        perror("No se puede crear el archivo de tiempo");
        exit(EXIT_FAILURE);
    }
    fprintf(file, "Tiempo de ejecución: %f segundos\n", time);
    fclose(file);
}

// Función para crear directorios de forma recursiva
void createDirectoriesRecursively(const char *path) {
    char temp[256];
    snprintf(temp, sizeof(temp), "%s", path);

    for (char *p = temp + 1; *p; p++) {
        if (*p == '/') {
            *p = '\0';
            if (mkdir(temp, 0777) == -1 && errno != EEXIST) {
                perror("Error creando directorio");
                exit(EXIT_FAILURE);
            }
            *p = '/';
        }
    }

    if (mkdir(temp, 0777) == -1 && errno != EEXIST) {
        perror("Error creando directorio");
        exit(EXIT_FAILURE);
    }
}

// Función principal
int main(int argc, char *argv[]) {
    if (argc < 2) {
        fprintf(stderr, "Uso: %s <archivo_matrices>\n", argv[0]);
        return EXIT_FAILURE;
    }

    const char *matricesFile = argv[1];

    long long ***matrices;
    int *rows, *cols, matrixCount;

    readMatrices(matricesFile, &matrices, &rows, &cols, &matrixCount);

    if (matrixCount < 2) {
        fprintf(stderr, "Error: Se requieren al menos dos matrices para multiplicar\n");
        return EXIT_FAILURE;
    }

    if (cols[0] != rows[1]) {
        fprintf(stderr, "Error: Las matrices no se pueden multiplicar\n");
        return EXIT_FAILURE;
    }

    long long **result = malloc(rows[0] * sizeof(long long *));
    for (int i = 0; i < rows[0]; i++) {
        result[i] = calloc(cols[1], sizeof(long long));
    }

    clock_t start = clock();
    multiplyMatricesWinograd(matrices[0], matrices[1], result, rows[0], cols[0], cols[1]);
    clock_t end = clock();

    char tiempoFile[150], resultadoFile[150];
    snprintf(tiempoFile, sizeof(tiempoFile), "../../../../Resultados/WinogradOriginal/c/tiempo/tiempo_WinogradOriginal_%dx%d.txt", rows[0], cols[1]);
    snprintf(resultadoFile, sizeof(resultadoFile), "../../../../Resultados/WinogradOriginal/c/resultadomultiplicacion/resultado_WinogradOriginal_%dx%d.txt", rows[0], cols[1]);

    createDirectoriesRecursively("../../../../Resultados/WinogradOriginal/c/tiempo");
    createDirectoriesRecursively("../../../../Resultados/WinogradOriginal/c/resultadomultiplicacion");

    double elapsedTime = (double)(end - start) / CLOCKS_PER_SEC;
    writeExecutionTime(tiempoFile, elapsedTime);
    writeMatrix(resultadoFile, result, rows[0], cols[1]);

    for (int i = 0; i < matrixCount; i++) {
        for (int j = 0; j < rows[i]; j++) free(matrices[i][j]);
        free(matrices[i]);
    }
    for (int i = 0; i < rows[0]; i++) free(result[i]);
    free(rows);
    free(cols);
    free(result);
    free(matrices);

    return EXIT_SUCCESS;
}

File: algoritmos/WinogradOriginal/src/python/WinogradOriginal.py
import os
import time

def multiply_matrices_winograd(a, b):
    row_a = len(a)
    col_a = len(a[0])
    col_b = len(b[0])

    row_factor = [0] * row_a
    col_factor = [0] * col_b

    for i in range(row_a):
        for j in range(col_a // 2):
            row_factor[i] += a[i][2 * j] * a[i][2 * j + 1]

    for i in range(col_b):
        for j in range(len(b) // 2):
            col_factor[i] += b[2 * j][i] * b[2 * j + 1][i]

    result = [[-row_factor[i] - col_factor[j] for j in range(col_b)] for i in range(row_a)]

    for i in range(row_a):
        for j in range(col_b):
            for k in range(col_a // 2):
                result[i][j] += (a[i][2 * k] + b[2 * k + 1][j]) * (a[i][2 * k + 1] + b[2 * k][j])

    if col_a % 2 == 1:
        for i in range(row_a):
            for j in range(col_b):
                result[i][j] += a[i][col_a - 1] * b[col_a - 1][j]

    return result


def read_matrices(filename):
    matrices = []
    with open(filename, 'r') as file:
        matrix = []
        for line in file:
            if line.strip():
                row = list(map(int, line.split()))
                matrix.append(row)
            else:
                if matrix:
                    matrices.append(matrix)
                    matrix = []
        if matrix:
            matrices.append(matrix)
    return matrices


def write_matrix(filename, matrix):
    with open(filename, 'w') as file:
        for row in matrix:
            file.write(' '.join(map(str, row)) + '\n')


def write_execution_time(filename, elapsed_time):
    with open(filename, 'w') as file:
        file.write(f'Tiempo de ejecución: {elapsed_time:.6f} segundos\n')


def create_directories_recursively(path):
    os.makedirs(path, exist_ok=True)


def main(matrices_file):
    matrices = read_matrices(matrices_file)

    if len(matrices) < 2:
        raise ValueError("Error: Se requieren al menos dos matrices para multiplicar")

    a = matrices[0]
    b = matrices[1]

    if len(a[0]) != len(b):
        raise ValueError("Error: Las matrices no se pueden multiplicar")

    start_time = time.time()
    result = multiply_matrices_winograd(a, b)
    elapsed_time = time.time() - start_time

    # Crear directorios para guardar resultados
    tiempo_file = "../../../../Resultados/WinogradOriginal/python/tiempo/tiempo_WinogradOriginal_{}x{}.txt".format(len(a), len(b[0]))
    resultado_file = "../../../../Resultados/WinogradOriginal/python/resultadomultiplicacion/resultado_WinogradOriginal_{}x{}.txt".format(len(a), len(b[0]))

    create_directories_recursively("../../../../Resultados/WinogradOriginal/python/tiempo")
    create_directories_recursively("../../../../Resultados/WinogradOriginal/python/resultadomultiplicacion")

    write_execution_time(tiempo_file, elapsed_time)
    write_matrix(resultado_file, result)


if __name__ == "__main__":
    import sys

    if len(sys.argv) < 2:
        print(f"Uso: {sys.argv[0]} <archivo_matrices>")
        sys.exit(1)

    matrices_file = sys.argv[1]
    try:
        main(matrices_file)
    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)

